### PR TITLE
feat(stepper): Add unit tests for stepper

### PR DESCRIPTION
**Pull request by g1shin**

### CODE REVIEW

**Comment by shlomiassaf** [comment]
@g1shin Been playing with the stepper branch.

Any particular reason for the label's index not being part of the whole label?

Instead of:
```html
<div class="mat-stepper-index">{{i + 1}}</div>
<div class="mat-stepper-label">
  <!-- If there is a label template, use it. -->
  <ng-container *ngIf="step.stepLabel" [ngTemplateOutlet]="step.stepLabel.template"></ng-container>
  <!-- It there is no label template, fall back to the text label. -->
  <div *ngIf="!step.stepLabel">{{step.label}}</div></div>
```

do it like:
```html
<div class="mat-stepper-label">
  <ng-container *ngIf="step.stepLabel" [ngTemplateOutlet]="step.stepLabel.template"></ng-container>
  <!-- It there is no label template, fall back to the text label. -->
  <div *ngIf="!step.stepLabel">
    <div class="mat-stepper-index">
      {{i + 1}}
    </div>
    {{step.label}}
  </div>
</div>
```

Forcing the index does not allow real customization of the index, for example if I want to do roman numerals or A. B. C

**Comment by shlomiassaf** [comment]
Another thing that pops out is not using `ViewEncapsulation.None` in the component.

This does not align with other components...

**Comment by mmalerba** [comment]
@g1shin needs rebase

**Comment by g1shin** [comment]
Rebase done; ready for review :+1:

File: src/lib/stepper/stepper.spec.ts [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+import {By} from '@angular/platform-browser';
+import {MdHorizontalStepper} from './stepper-horizontal';
+import {MdVerticalStepper} from './stepper-vertical';
+import {NoopAnimationsModule} from '@angular/platform-browser/animations';
+import {FormControl, FormGroup, FormsModule, ReactiveFormsModule, Validators} from '@angular/forms';
+import {MdStepperNext, MdStepperPrevious} from './stepper-button';
+import {dispatchKeyboardEvent} from '@angular/cdk/testing';
+import {ENTER, LEFT_ARROW, RIGHT_ARROW, SPACE} from '@angular/cdk/keycodes';
+
+fdescribe('MdHorizontalStepper', () => {
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: Looks like you left an `fdescribe` by mistake. Remove?

File: src/lib/stepper/stepper.spec.ts [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [MdStepperModule, NoopAnimationsModule, FormsModule, ReactiveFormsModule],
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: Do you need `FormsModule`? It looks like you aren't using any template-driven directives.  Same on line 114.

File: src/lib/stepper/stepper.spec.ts [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+      stepperComponent = fixture.debugElement
+          .query(By.css('md-horizontal-stepper')).componentInstance;
+    });
+
+    it('should default to the first step', () => {
+      expect(stepperComponent.selectedIndex).toBe(0);
+    });
+
+    it('should change selected index on header click', () => {
+      let stepHeader = fixture.debugElement.queryAll(By.css('.mat-horizontal-stepper-header'));
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: Nit: Maybe name this `stepHeaders` (plural)? To make clear that this grabs headers from all steps. Same on line 142.

File: src/lib/stepper/stepper.spec.ts [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+                              fixture: ComponentFixture<any>,
+                              stepContent: DebugElement[]) {
+  let stepContentEl = stepContent[0].nativeElement;
+  expect(stepContentEl.getAttribute('aria-expanded')).toBe('true');
+
+  stepperComponent.selectedIndex = 1;
+  fixture.detectChanges();
+
+  expect(stepContentEl.getAttribute('aria-expanded')).toBe('false');
+  stepContentEl = stepContent[1].nativeElement;
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: Nit: It's a bit hard to follow when you're naming two different step elements with the same name. Maybe create a new `let` for this - `secondStepEl`?

File: src/lib/stepper/stepper.spec.ts [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+      checkSelectionChangeOnHeaderClick(stepperComponent, fixture, stepHeader);
+
+    });
+
+    it('should set the "tablist" role on stepper', () => {
+      stepperEl = fixture.debugElement.query(By.css('md-horizontal-stepper')).nativeElement;
+      expect(stepperEl.getAttribute('role')).toBe('tablist');
+    });
+
+    it('should expand the right content', () => {
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: Nit: In this case, you mean set `aria-expanded` correctly, right? I'd mention that in the test name. Same on line 152.

File: src/lib/stepper/stepper.spec.ts [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+          .query(By.css('md-horizontal-stepper')).componentInstance;
+    });
+
+    it('should have true linear attribute', () => {
+      expect(stepperComponent.linear).toBe(true);
+    });
+
+    it('should not move to next step if current step is not valid', () => {
+      expect(testComponent.oneGroup.get('oneCtrl')!.value).toBe('');
+      expect(testComponent.oneGroup.valid).toBe(false);
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: Nit: Here I'd expect a test that the control itself is invalid (if you were to add another control, it could cause the parent form to be false for a different reason).  Same on 199.

File: src/lib/stepper/stepper.spec.ts [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+      expect(stepperComponent.selectedIndex).toBe(0);
+
+      let stepHeaderEl = fixture.debugElement
+          .queryAll(By.css('.mat-horizontal-stepper-header'))[1].nativeElement;
+      checkLinearStepperValidity(stepHeaderEl, stepperComponent, testComponent, fixture);
+    });
+  });
+});
+
+fdescribe('MdVerticalStepper', () => {
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: fdescribe

File: src/lib/stepper/stepper.spec.ts [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+                            fixture: ComponentFixture<any>,
+                            stepHeader: DebugElement[]) {
+  expect(stepperComponent._focusIndex).toBe(0);
+  expect(stepperComponent.selectedIndex).toBe(0);
+
+  let stepHeaderEl = stepHeader[0].nativeElement;
+  dispatchKeyboardEvent(stepHeaderEl, 'keydown', RIGHT_ARROW);
+  fixture.detectChanges();
+
+  expect(stepperComponent._focusIndex).toBe(1);
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: I think fallback messages for these would be helpful to guide us through your thinking.

File: src/lib/stepper/stepper.spec.ts [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+  inputLabel = 'Step 3';
+}
+
+@Component({
+  template: `
+    <md-vertical-stepper linear>
+      <md-step [stepControl]="oneGroup">
+        <form [formGroup]="oneGroup">
+          <ng-template mdStepLabel>Step one</ng-template>
+          <md-input-container>
```
Comment (kara) [inline src/lib/stepper/stepper.spec.ts]: Do you want to use md-form-field on these to be forward-friendly? cc @mmalerba

**Comment by g1shin** [comment]
Changes have been made based on review. Ready for review again :smiley:

**Comment by kara** [comment]
@mmalerba Do you want to take a look before I apply merge labels?

File: src/lib/stepper/stepper.spec.ts [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+      ]
+    });
+
+    TestBed.compileComponents();
+  }));
+
+  describe('basic horizontal stepper', () => {
+    let fixture: ComponentFixture<SimpleMdHorizontalStepperApp>;
+    let stepperComponent: MdHorizontalStepper;
+    let stepperEl: HTMLElement;
```
Comment (mmalerba) [inline src/lib/stepper/stepper.spec.ts]: since only 1 test uses this you can just declare it in that test

File: src/lib/stepper/stepper.spec.ts [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+      let stepHeaderEl = fixture.debugElement
+          .queryAll(By.css('.mat-vertical-stepper-header'))[1].nativeElement;
+
+      checkLinearStepperValidity(stepHeaderEl, stepperComponent, testComponent, fixture);
+    });
+  });
+});
+
+function checkSelectionChangeOnHeaderClick(stepperComponent:
+                                               MdHorizontalStepper | MdVerticalStepper,
```
Comment (mmalerba) [inline src/lib/stepper/stepper.spec.ts]: just use common base class MdStepper? (here and other functions that take a stepper)

File: src/lib/stepper/stepper.spec.ts (hunk 9 above)
Comment (mmalerba) [inline src/lib/stepper/stepper.spec.ts]: +1 if you're synced up to where md-form-field exists

**Comment by g1shin** [comment]
Ready for review again

File: src/lib/stepper/stepper.spec.ts [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+    });
+
+    it('should default to the first step', () => {
+      expect(stepperComponent.selectedIndex).toBe(0);
+    });
+
+    it('should change selected index on header click', () => {
+      let stepHeaders = fixture.debugElement.queryAll(By.css('.mat-horizontal-stepper-header'));
+      checkSelectionChangeOnHeaderClick(stepperComponent, fixture, stepHeaders);
+
```
Comment (mmalerba) [inline src/lib/stepper/stepper.spec.ts]: nit: remove extra line
